compiler/next/lib/uast/For.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chpl/uast/For.h"
+
+#include "chpl/uast/Builder.h"
+
+namespace chpl {
+namespace uast {
+
+
+static int32_t computeStatementChildNum(int8_t indexVarChildNum,
+                                        int8_t iterandChildNum) {
+  return (indexVarChildNum >= 0) + (iterandChildNum >= 0);
+}
+
+For::For(ASTList children, int8_t indexVarChildNum, int8_t iterandChildNum,
+         bool usesDo,
+         bool expressionLevel,
+         bool param)
+  : Loop(asttags::For, std::move(children),
+         computeStatementChildNum(indexVarChildNum, iterandChildNum),
+         usesDo),
+    indexVarChildNum_(indexVarChildNum),
+    iterandChildNum_(iterandChildNum),
+    expressionLevel_(expressionLevel),
+    param_(param) {
+
+#ifndef NDEBUG
+  assert(iterandChildNum >= 0);
+  // check that all children are exprs (and not, say, Symbols)
+  for (const ASTNode* child : this->children()) {
+    assert(child->isExpression());
+  }
+#endif
+}
```

Comment:
Let's move the constructor to the header (see above comment). Also, replace the loop with `    assert(isExpressionASTList(children_));`.